packages/protocol/contracts/L1/ProverPool.sol

```diff
@@ -0,0 +1,479 @@
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.20;
+
+import { AddressResolver } from "../common/AddressResolver.sol";
+import { EssentialContract } from "../common/EssentialContract.sol";
+import { TaikoToken } from "./TaikoToken.sol";
+import { Proxied } from "../common/Proxied.sol";
+
+interface IProverPool {
+    function assignProver(
+        uint64 blockId,
+        uint32 feePerGas
+    )
+        external
+        returns (address prover, uint32 rewardPerGas);
+
+    function releaseProver(address prover) external;
+    function slashProver(address prover) external;
+
+    // TODO: we need to add this as well.
+    function getAvailableCapacity() external returns (uint256);
+}
+
+contract ProverPool is EssentialContract, IProverPool {
+    // New concept TLDR:
+    // 2 main data 'registry':
+    // - TopProver array: we keep track of the top32. We need an array because we need to iterate over
+    // - ExitingProver mapping: each and every prover who is either exiting or kicked out (by someone staking more) is
+    // here until they are not withdraw the funds (with exit())
+
+    // NOTE:
+    // - signalling an exit if you are a prover can be done via stake(totalAMount = 0)
+    // - modifying existing parameters also done via stake (see comments there)
+    // - there are mechanisms in the stake() code to handle scenarios like:
+    //     A: Bob is the least staker. Alice comes in and stakes more than Bob and Carol.
+    //     Now Bob forced to leave and not within the ExitingProvers. He realizes it and
+    //     tries to re-stake. Now he stakes more than Carol, so Carol is gone, but Bob can
+    //     re-use his funds currently in the 'staking queue'. So basically he is out of the
+    //     exiting queue.
+    //
+    //     B: Almost same situation as above, except that when Bob is leaving (force leaving)
+    //     Carol decided to leave as well, so she is not in the topProver array anymore. It means
+    //     Bob could come back with his stake (currenlty in the exit 'queue') or even less than that
+    //     amount since Carol left, there is 1 "free space". So he could decide to re-stake less than
+    //     his amount in the current exit 'queue', so technically he still have some left in the exit queue
+    //     while also being a topProver in the same time!
+
+
+    // This is 1/4 slot
+    // TODO: Dani, this structure should hold all information required to
+    // dynamically calculate the weight of the top 32 provers and select one
+    // based on a random number.
+    struct TopProver {
+        uint32 amount; // this value will change when we slash the prover
+        uint16 rewardPerGas;
+        uint16 currentCapacity;
+    }
+
+    // prover always have one Prover object,regardless if it is active or exiting.
+    // @TODO: We do not need to use both. We shall have 1 for TopProvers, and 1 for ExitingProvers
+    struct ExitingProver {
+        uint64 exitRequestedAt;
```

Comment:
If we name this structure as `ExitingProver`, we can remove the 'exit" prefix from its variables.